engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/state/PutStateNode.java

```diff
@@ -1,13 +1,72 @@
 package org.enso.interpreter.node.expression.builtin.state;
 
+import com.oracle.truffle.api.TruffleLanguage;
+import com.oracle.truffle.api.dsl.*;
 import com.oracle.truffle.api.nodes.Node;
+import org.enso.interpreter.Language;
 import org.enso.interpreter.dsl.BuiltinMethod;
 import org.enso.interpreter.dsl.MonadicState;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.state.data.SingletonMap;
+import org.enso.interpreter.runtime.state.data.SmallMap;
+import org.enso.interpreter.runtime.error.PanicException;
 import org.enso.interpreter.runtime.state.Stateful;
 
 @BuiltinMethod(type = "State", name = "put", description = "Updates the value of monadic state.")
-public class PutStateNode extends Node {
-  Stateful execute(@MonadicState Object state, Object _this, Object new_state) {
-    return new Stateful(new_state, state);
+@ImportStatic(SmallMap.class)
+@ReportPolymorphism
+public abstract class PutStateNode extends Node {
+  static PutStateNode build() {
+    return PutStateNodeGen.create();
+  }
+
+  abstract Stateful execute(@MonadicState Object state, Object _this, Object key, Object new_state);
+
+  @Specialization(guards = "state.getKey() == key")
+  Stateful doExistingSingleton(SingletonMap state, Object _this, Object key, Object new_state) {
+    return new Stateful(new SingletonMap(key, new_state), new_state);
+  }
+
+  @Specialization(
+      guards = {"state.getKeys() == cachedKeys", "index != NOT_FOUND", "key == cachedKey"})
+  Stateful doExistingMultiCached(
+      SmallMap state,
+      Object _this,
+      Object key,
+      Object new_state,
+      @Cached("key") Object cachedKey,
+      @Cached(value = "state.getKeys()", dimensions = 1) Object[] cachedKeys,
+      @Cached("state.indexOf(key)") int index) {
+    Object[] newVals = new Object[cachedKeys.length];
+    System.arraycopy(state.getValues(), 0, newVals, 0, cachedKeys.length);
+    newVals[index] = new_state;
+    SmallMap newState = new SmallMap(cachedKeys, newVals);
+    return new Stateful(newState, new_state);
```

Comment:
Having `newState` and `new_state` on the same line, meaning different things is extremely confusing. Maybe better names could be found here?